pkg/kubelet/kubelet.go

```diff
@@ -2076,6 +2077,15 @@ func (kl *Kubelet) SyncPod(ctx context.Context, updateType kubetypes.SyncPodType
 				kl.statusManager.SetPodResizeInProgressCondition(pod.UID, v1.PodReasonError, r.Message, false)
 			}
 		}
+
+		if errors.Is(r.Error, images.ErrImagePull) {
+			hasPullFailures = true
+		}
+	}
+
+	// If we have any image pull errors, report any failed pull secrets
+	if hasPullFailures && len(failedPullSecretNames) > 0 {
+		kl.recorder.Eventf(pod, v1.EventTypeWarning, "FailedToRetrieveImagePullSecret", "Unable to retrieve some image pull secrets: %v", failedPullSecretNames)
```

Comment:
I wonder if there may be some unfortunate situation when we enter the sync pod many times and it always returns the `ErrImagePull` on one of the containers, while the reason for the loop is something else. Effectively making us write this message way more times than expected? One way to try is to make one of the containers enter the image pull backoff - waiting fro 5 minutes while other containers keep restarting for some other reason.

I didn't look into code too deep at this moment. So may be it can be explained that this will not happen.

Reply:
If a sync occurs and a container is already failing to pull, it should be in backoff, so the pull won't be attempted, and the sync result for that container will be `ErrImagePullBackoff`: https://github.com/kubernetes/kubernetes/blob/d202031fe071eb5500decb2e61e8cb7333635231/pkg/kubelet/images/image_manager.go#L279

The warning will only trigger when a pull is attempted that results in `ErrImagePull` (so `ErrImagePullBackoff` won't trigger it), and a pull secret is missing: https://github.com/kubernetes/kubernetes/blob/f5a181e9e0df9ed4da9dc60723c1ff0c625ca872/pkg/kubelet/kubelet.go#L2058

I think that's the correct behaviour, and is certainly no worse than the current behaviour of emitting the warning on every sync when a secret is missing.